### ISSUE TITLE
Difficulty filter does not respect modifiers difficulty change

### ISSUE THREAD

**Issue by stevefan1999-personal** [ppy/osu-stable-issues#712]
# Synopsis 
Video evidence should worth a thousand words:

https://user-images.githubusercontent.com/29133953/113407902-dd519200-93e0-11eb-9509-45688b893dc5.mp4

# Explanation ~(for the idiots who still insist words are better than video)~
1. The search filter is applied before using any modifiers
2. Because we've already filtered the maps, then they will not be considered to be calculated difficulty at all (for example, Linada's map had 3.89 stars only, but it is less than the 3.9 I set)
3. Then when we applied modifiers, the state of the stars changed (now Linada's map is "updated" to 5.39 stars), but because that map does not meet the criteria to be included in the selection menu (or again in other word, filtered apriori due to it only have 3.89 stars), so even though the modifier had the adjusted difficulty matches the filter range (from 3.89 to 5.39), it won't appear in the search at all

# Potential fix
This is probably unfixable unless the metadata state (such as BPM/Stars/CS) can be refreshed dynamically, but this also suggested to use a dynamic data structure, which means it will have substantial runtime impact for low-end machine, or for people who have more than 1000 beatmaps

Another way of fixing it:
1. Keep the original result set
 1.1 This also implies you will have the filtered set (for example, Filtered = All Songs - filter(Star Range, Result))
2. Because the result set is always going to fit in the range regardless of the strong modifiers, we can just search for all the difficulties in the filtered set after update
3. Then insert the missing map from the filtered map **ordered**
 3.1 This means something like a priority queue is needed
4. If the relation is inverted (i.e. using easy mod), than do the same but flip the result and filtered set

Here is the list time/space complexity of priority queue implementations:
![image](https://user-images.githubusercontent.com/29133953/113410328-bcd80680-93e5-11eb-80c2-e9664fc523f3.png)

It's better to choose implementation that's better than O(n) in melding (~haha no binary heap bitches~), because technically it also mean queue combination/union and its handy in this case, there are plenty of Fibonacci priority queue implementation, ~though I'm not certain about any implementations in .net/C#~ [Here is a repo with source code for you to copy paste peppy](https://github.com/justcoding121/advanced-algorithms/blob/develop/src/Advanced.Algorithms/DataStructures/Heap/FibonacciHeap.cs).

# Disclaimer

Okay, I can see this coming that some ~idiots~ people will try to laugh at me saying why are you playing a PP map, but I just see this phenomenon and I have reported it truly in details, please respect the technicality rather than trying to judge me from ethical/moral standpoint, this is a warning so you would better STFU.

**Comment by smoogipoo** [comment]
The structure in stable makes this a little too complicated, so I'm marking this as out of scope. This will eventually work in lazer, but is held up by a few other things right now.

**Comment by stevefan1999-personal** [comment]
bruh
